proxy/coco/src/peer.rs

```diff
@@ -71,100 +100,235 @@ impl Peer {
+        let state = state.clone();
+        let store = store.clone();
+        let notifier = announcer_notifier.clone();
+        tokio::spawn(future::Abortable::new(
+            async move {
+                loop {
+                    log::trace!("announcer: waiting to get notified");
+                    notifier.notified().await;
+                    log::trace!("announcer: notified");
+                    announce(&state, &store, &mut announce_sender).await
+                }
+            },
+            abort_registration,
+        ));
+
+        abort_handle
+    };
+
+    let (mut peer_sync_queue, mut peer_sync_receiver) = mpsc::channel::<PeerId>(RECEIVER_CAPACITY);
+    let peer_sync_handle = {
+        let (abort_handle, abort_registration) = future::AbortHandle::new_pair();
+        let state = state.clone();
+        tokio::spawn(future::Abortable::new(
+            async move {
+                while let Some(peer_id) = peer_sync_receiver.recv().await {
+                    log::trace!("syncer: received peer id {}", peer_id);
+                    sync(&state, peer_id, &mut peer_sync_sender).await;
+                }
+            },
+            abort_registration,
+        ));
+
+        abort_handle
+    };
+
+    loop {
+        let event = tokio::select! {
+            _ = announce_timer.tick() => Event::Announce(AnnounceEvent::Tick),
+            Some(announce_event) = announcements.recv() => Event::Announce(announce_event),
+            Some(protocol_event) = protocol_events.next() => Event::Protocol(protocol_event),
+            Some(sync_event) = peer_syncs.recv() => Event::PeerSync(sync_event),
+            Some(timeout_event) = timeouts.recv() => {
+                peer_sync_handle.abort();
+                Event::Timeout(timeout_event)
+            },
+            else => {
+                break
+            },
+        };
+
+        // Send will error if there are no active receivers. This
+        // case is expected and should not terminate the run loop.
+        subscriber.send(event.clone()).ok();
+        log::debug!("{:?}", event);
+
+        for cmd in run_state.transition(event) {
+            match cmd {
+                Command::Announce => announcer_notifier.notify(),
+                Command::SyncPeer(peer_id) => {
+                    peer_sync_queue.send(peer_id).await.ok();
+                },
+                Command::StartSyncTimeout(sync_period) => {
+                    tokio::spawn(start_sync_timeout(sync_period, timeout_sender.clone()));
                 },
             };
-        });
+        }
     }
 
-    /// Sync timeout subroutine.
-    fn start_sync_timeout(sync_period: Duration, sender: mpsc::Sender<TimeoutEvent>) {
-        tokio::spawn(async move {
-            tokio::time::delay_for(sync_period).await;
-            sender.clone().send(TimeoutEvent::SyncPeriod).await.ok();
-        });
-    }
+    announcer_handle.abort();
```

Comment:
Not entirely sure about this, actually. Could be we need to wake the tasks when the parent gets aborted.